File: Code/C++/stock_span.cpp
/*
    In this problem we are provided with the stock prices of various days and we have to find
    the span of each day. The span of ith day is the total number of consecutive days 
    before it that had stock prices strictly less than the ith day. the minimum span of a day is 1.
    For Ex:
           Input: 
                  5 
                  100 10 80 29 4
           Output: 
                  1 1 2 1 1
*/
#include<bits/stdc++.h>
using namespace std;
int main()
{
  //Taking input
  int n;
  cout<<"Enter number of days\n";
  cin>>n;
  int a[n],span[n];
  for(int i=0;i<n;i++)
  {
    cout<<"Enter price of stock on day "<<i<<"\n";
    cin>>a[i];
  }
  
  //Creating a stack to calculate span of each day
  stack<int> s;
  
  for(int i=0;i<n;i++)
  {
    
    //This pops out the elements that are smaller than a[i] out of the stack
     while((!s.empty()) && a[s.top()]<=a[i])
     {
       s.pop();
     }
    
    // If the stack this empty it means that a[i] is the largest value
    //So its span is i+1
    if(s.empty())
    {
      span[i]=i+1;
    }
    else
    {
      span[i]=i-s.top();
    }
    
    //In the end we push a[i] in the stack to compare it with other elements
    s.push(i);
  }
  
  //Printing the span of each day
  for(int i=0;i<n;i++)
  {
    cout<<span[i]<<" ";
  }
  
}

/*      TEST CASES

    1) INPUT:
               7
               100 20 40 30 5 101 70
       OUTPUT:
                1 1 2 1 1 6 1
    
    2) INPUT:
               6
               10 200 40 39 51 101 
       OUTPUT:
                1 2 1 1 3 4 
                
*/
           

File: Code/Python/StockSpan.py
'''

Problem: User would be given with the prices of stock.User needs to output the span of each stock price for n day,where 
span refers to as  maximum number of consecutive days  for which the price of the stock on the current day is less than or equal to the price on the given day.

Solution:
The problem could be done in O(n) time complexity using stack.The span of stock at day 1 would always be 1 as there are no previous day before it.
Index of the first day is pushed inside the stack,if price at the current day is less than the price of stack that is stored at stack top 
then span of the current day would be currentday-stack_top,after which we would push current day index in the stack.
If price at current day is greater than price at the index stored by stack top then we would simply pop from the stack,
untill stack become empty or price at stack_top is greater than cureent day price.
If stack become empty then span at that day would be current_day+1

'''
stock = []
n = int(input("Enter total number of days : "))
for i in range(0, n):
	t = int(input())

	stock.append(t)
answer=[]
for i in range(0,n):
    answer.append(0)

answer[0]=1
stackhelp=[]
stackhelp.append(0)
for i in range(1,n):
    if(len(stackhelp)>0 and stock[i]<stock[stackhelp[-1]]):
        answer[i]=i-stackhelp[-1]
    else:
        while(len(stackhelp)>0 and stock[i]>=stock[stackhelp[-1]]):
            stackhelp.pop()
        if(len(stackhelp)<=0):
            answer[i]=i+1;
        else:
            answer[i]=i-stackhelp[-1]
    stackhelp.append(i)
print("Resultant span of each stock is : ")
for i in range(0, n):
		print (answer[i], end =" ")

'''
Time Complexity : O(N)
Space Complexity: O(N)

Test Cases:

Input 1:
5
30 35 40 38 35

Output 1:
1 2 3 1 1


Input 2:
7
1 2 3 8 5 6 7

Output 2:
1 2 3 4 1 2 3

'''